Document kubeadm OS version requirement (#50375)
kubernetes/website: content/en/docs/setup/production-environment/tools/kubeadm/install-kubeadm.md

<!-- steps -->

## Check your OS kernel version

> 

Reviewer: on this line we need to add `{{% thirdparty-content %}}` if we link to external pages outside of k8.io

Author: Fixed in https://github.com/kubernetes/website/pull/50375/commits/66b032bdbdd12d84aee7ed3e95bbe4586147f230

> {{< tabs name="container_runtime" >}}

Reviewer: should be just `OS`?

Reviewer: i think i don't understand how these tabs work, as it currently says `{{< tabs name="container_runtime" >}}` and i though it should be something like `{{< tabs name="OS" >}}`, indicating that the tabs in this widget are about Linux/Windows OS.

Reviewer 2: ```suggestion
{{< tabs name="operating_system_version_check" >}}
```

Author: The suggestion makes sense. I thought the original comment was about the heading, not the  tab name.

{{% tab name="Linux" %}}

> * The Linux kernel must be 4.19 or newer, 5.x, or 6.x.

Reviewer: i think we should link to the external wiki page, like i said on the ticket.
we will keep system-validators up to date with lts releases, but i don't want us to have to also update k/website about this all the time.

linking to the code of system-vaidators is not ideal either.

Reviewer 3: We can link to this page that maintains the kernel versions: https://www.kernel.org/category/releases.html

The 4.19 kernel version is already deprecated and support for it is removed in the validators code.

Reviewer: there is an open pr to update system-validators in k/k, it missied code freeze.
- https://github.com/kubernetes/kubernetes/pull/130804

> The source of truth is the system-validators code used by kubeadm, i.e. the code vendored in kubernetes/kubernetes.

i wouldn't mind linking to the source code, but sig-docs have not accepted that historically.

> Those external sources are likely to be out of sync with actual kubeadm requirements. But any explictly mentioned versions will also go out of sync, unless we frequently update this page.

i think we can just link to external official docs but ease the language. ala 'kubeadm tries to keep up with supporting only lts kernels'

> Can kubeadm today print out the required versions? If not, we could add a command, and then reference the command in this page, e.g. "For the required versions, run kubeadm ..." Actually, the best kubeadm could do right now is print a regular expression. That's not too helpful.

it can but i don't want us to add additional commands. the user can call init/join with --dry-run and it will report the system errors.

Reviewer 3: Going back to the original point -- the kernel website is the best place to figure out what versions are LTS.

> * You can get the kernel version using the command `uname -r`

Reviewer 3: this is slightly confusing -- this command should be run on the machine that will be running `kubeadm` but that's not clear here.

Author: This page is about installing kubeadm. I find it hard to imagine any commands should be run on a different machine than the one where you are installing kubeadm.

(I used the same wording as another section of this page, https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#verify-mac-address.)

Reviewer 3: fair enough, if there's already precedence then its good!

{{% /tab %}}

{{% tab name="Windows" %}}

* The Windows kernel version must be 10.0.14393 or newer.

> * You can get the kernel version (also called the OS version) using the command `systeminfo`

Reviewer: a similar case here.

we can link to these 2 pages:
- https://learn.microsoft.com/en-us/windows/release-health/windows-server-release-info
- https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/#before-you-begin
instead of hardcoding the version.

Author: Fxied

{{% /tab %}}
{{< /tabs >}}

A Kubernetes cluster created by kubeadm depends on software that use kernel features.
This software includes, but is not limited to, the container runtime (such as containerd), kubelet, kube-proxy, and a CNI plugin.
The software depends on specific kernel features, but does not typically publish a list of these features. Often, the software requires a range of kernel versions.
To make sure this software runs successfully, kubeadm also requires a range of kernel versions.

{{< note >}}

To help you avoid unexpected errors as a result of an unsupported kernel version, kubeadm runs the `SystemVerification` pre-flight check.
This check fails if the kernel version is not supported. You may choose to skip the check, if you know that your kernel provides the required features, even though kubeadm does not support its version.

> {{< /note >}}

Reviewer: can be a regular paragraph outside of a note.

Author: Fixed in https://github.com/kubernetes/website/pull/50375/commits/66b032bdbdd12d84aee7ed3e95bbe4586147f230

<!-- steps -->

## Check your OS version

{{% thirdparty-content %}}

{{< tabs name="container_runtime" >}}
{{% tab name="Linux" %}}

* The kubeadm project supports LTS kernels. See [List of LTS kernels](https://www.kernel.org/category/releases.html).
* You can get the kernel version using the command `uname -r`

{{% /tab %}}

{{% tab name="Windows" %}}

> * The kubeadm project supports recent kernel versions. For a list recent kernels, see [Windows Server Release Information](https://learn.microsoft.com/en-us/windows/release-health/windows-server-release-info).

Reviewer 2: ```suggestion
* The kubeadm project supports recent kernel versions. For a list of recent kernels, see [Windows Server Release Information](https://learn.microsoft.com/en-us/windows/release-health/windows-server-release-info).
```

* You can get the kernel version (also called the OS version) using the command `systeminfo`

For more information, see See [Adding Windows Nodes](/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/#before-you-begin).

{{% /tab %}}
{{< /tabs >}}

A Kubernetes cluster created by kubeadm depends on software that use kernel features.

> This software includes, but is not limited to, the container runtime (such as containerd), kubelet, kube-proxy, and a CNI plugin.

Reviewer 2: ```suggestion
This software includes, but is not limited to, the
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}, the kubelet, 
and a network (CNI) plugin.
```

We avoid suggesting one product and not another. I also dropped kube-proxy because although it relies on a kernel (doesn't all userland?), it's optional.

> The software depends on specific kernel features, but does not typically publish a list of these features. Often, the software requires a range of kernel versions.

Reviewer 2: Is this what we're saying here?
```diff
-The software depends on specific kernel features, but does not typically publish a list of these features. Often, the software requires a range of kernel versions.
+Kubernetes components can depend on specific kernel features, but the Kubernetes project
+typically does **not** publish a list of the required kernel features.
```

I guess we don't document this because of capacity, right? Not because of other motivations.

Author: I see your motivation in limiting the statement to "Kubernetes components," but I simply can't say that it is the Kubernetes components that kubeadm is concerned about. My impression was that it was third-party components like the container runtime, and the CNI plugin. @neolit123, WDYT?

Reviewer 4: Or we may link to https://kubernetes.io/docs/reference/node/kernel-version-requirements/ or something like this?

Reviewer: > I see your motivation in limiting the statement to "Kubernetes components," but I simply can't say that it is the Kubernetes components that kubeadm is concerned about. My impression was that it was third-party components like the container runtime, and the CNI plugin. @neolit123, WDYT?

i think we should be broad. both k8s components and external deps can depend on kernel features.

Author: > both k8s components and external deps can depend on kernel features.

It really does not make sense to me that kubeadm care about external dependencies (i.e. dependencies installed by the user, not kubeadm) like the container runtime and the CNI plugin.

If it were up to me, I would keep @lmktfy's proposed wording, and I would add to it the link @pacoxu suggested above.

However, I will defer to you.

Reviewer: the proposed https://github.com/kubernetes/website/pull/50375#discussion_r2033148607 sgtm
